File: OEM/Domain/ComplementaryTasks/ComplementaryTask.cs
using System;
using DDDSample1.Domain.Shared;
using OEMAPI.Domain.ComplementaryTaskCategories;

namespace OEMAPI.Domain.ComplementaryTasks
{
    public class ComplementaryTask : Entity<ComplementaryTaskId>, IAggregateRoot
    {
        public ComplementaryTaskCategory Category { get; private set; }
        public VveId VveId { get; private set; }
        public ResponsibleTeam ResponsibleTeam { get; private set; }
        public StartTime StartTime { get; private set; }
        public EndTime EndTime { get; private set; }
        public Status Status { get; private set; }

        private ComplementaryTask() { }

        public ComplementaryTask(
            ComplementaryTaskCategory category,
            VveId vveId,
            ResponsibleTeam responsibleTeam,
            StartTime startTime)
        {
            Id = new ComplementaryTaskId(Guid.NewGuid().ToString());
            Category = category;
            VveId = vveId;
            ResponsibleTeam = responsibleTeam;
            StartTime = startTime;
            EndTime = new EndTime(null);   // ongoing task
            Status = Status.Ongoing();
        }

        public void Complete(EndTime endTime)
        {
            if (endTime.Value < StartTime.Value)
                throw new BusinessRuleValidationException("End time cannot be before start time.");

            EndTime = endTime;
            Status = Status.Completed();
        }
    }
}

File: OEM/src/Domain/ComplementaryTasks/ComplementaryTask.ts
import { AggregateRoot } from "../../core/domain/AggregateRoot";
import { UniqueEntityID } from "../../core/domain/UniqueEntityID";
import { Result } from "../../core/logic/Result";
import { Guard } from "../../core/logic/Guard";

import { ComplementaryTaskId } from "./ComplementaryTaskId";
import { VesselVisitExecutionId } from "../VesselVisitExecutions/VesselVisitExecutionId";
import { ResponsibleTeam } from "./ResponsibleTeam";
import { StartTime } from "./StartTime";
import { EndTime } from "./EndTime";
import { ComplementaryTaskStatus } from "./ComplementaryTaskStatus";
import { ComplementaryTaskCategoryId } from "../ComplementaryTaskCategories/ComplementaryTaskCategoryId";

interface ComplementaryTaskProps {
  vesselVisitExecutionId: VesselVisitExecutionId;
  categoryId: ComplementaryTaskCategoryId;
  responsibleTeam: ResponsibleTeam;
  startTime: StartTime;
  endTime?: EndTime;
  status: ComplementaryTaskStatus;
}

export class ComplementaryTask extends AggregateRoot<ComplementaryTaskProps> {

  get id(): UniqueEntityID {
    return this._id;
  }

  get taskId(): ComplementaryTaskId {
    return ComplementaryTaskId.create(this.id);
  }

  get vesselVisitExecutionId(): VesselVisitExecutionId {
    return this.props.vesselVisitExecutionId;
  }

  get categoryId(): ComplementaryTaskCategoryId {
    return this.props.categoryId;
  }

  get responsibleTeam(): ResponsibleTeam {
    return this.props.responsibleTeam;
  }

  get startTime(): StartTime {
    return this.props.startTime;
  }

  get endTime(): EndTime | undefined {
    return this.props.endTime;
  }

  get status(): ComplementaryTaskStatus {
    return this.props.status;
  }

  private constructor(props: ComplementaryTaskProps, id?: UniqueEntityID) {
    super(props, id);
  }

  public static create(
    props: ComplementaryTaskProps,
    id?: UniqueEntityID
  ): Result<ComplementaryTask> {

    const guardedProps = [
      { argument: props.vesselVisitExecutionId, argumentName: "vesselVisitExecutionId" },
      { argument: props.categoryId, argumentName: "categoryId" },
      { argument: props.responsibleTeam, argumentName: "responsibleTeam" },
      { argument: props.startTime, argumentName: "startTime" },
      { argument: props.status, argumentName: "status" }
    ];

    const guardResult = Guard.againstNullOrUndefinedBulk(guardedProps);

    if (!guardResult.succeeded) {
      return Result.fail<ComplementaryTask>(guardResult.message);
    }

    const task = new ComplementaryTask({ ...props }, id);
    return Result.ok<ComplementaryTask>(task);
  }
}
